File: ml/src/contracts.py
import math
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class Contract(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Landmark2D(Contract):
    id: int = Field(ge=0, le=1_024)
    x: float
    y: float
    confidence: float = Field(ge=0.0, le=1.0)

    @model_validator(mode="after")
    def finite(self) -> "Landmark2D":
        if not math.isfinite(self.x) or not math.isfinite(self.y):
            raise ValueError("landmark coordinates must be finite")
        return self


class CameraObservation(Contract):
    device_id: str = Field(min_length=3, max_length=128)
    frame_id: int = Field(ge=0)
    captured_at_ms: int = Field(ge=0)
    image_width: int = Field(ge=1, le=16_384)
    image_height: int = Field(ge=1, le=16_384)
    rotation_degrees: Literal[0, 90, 180, 270] = 0
    mirrored: bool = False
    track_id: int = Field(default=1, ge=0)
    landmarks: list[Landmark2D] = Field(min_length=1, max_length=133)

    @model_validator(mode="after")
    def unique_landmarks(self) -> "CameraObservation":
        ids = [landmark.id for landmark in self.landmarks]
        if len(ids) != len(set(ids)):
            raise ValueError("landmark IDs must be unique per camera observation")
        return self


class ReconstructionRequest(Contract):
    schema_version: Literal[1] = 1
    session_id: str = Field(min_length=3, max_length=128)
    exercise: str = Field(default="left_bicep_curl", min_length=1, max_length=64)
    observations: list[CameraObservation] = Field(min_length=2, max_length=8)

    @model_validator(mode="after")
    def unique_devices(self) -> "ReconstructionRequest":
        devices = [observation.device_id for observation in self.observations]
        if len(devices) != len(set(devices)):
            raise ValueError("one synchronized observation per device is required")
        return self


class Joint3D(Contract):
    x: float
    y: float
    z: float
    confidence: float = Field(ge=0.0, le=1.0)
    observations: int = Field(ge=2)


class AngleMetric(Contract):
    name: str
    degrees: float
    joint_ids: tuple[int, int, int]


class ResultMetadata(Contract):
    coordinate_system: Literal["camera_a_world_right_handed"] = "camera_a_world_right_handed"
    units: str
    calibration_id: str
    reprojection_error: float | None
    source_frame_ids: dict[str, int]
    source_timestamps_ms: dict[str, int]
    pairing_delta_ms: int
    processing_time_ms: float


class ReconstructionResult(Contract):
    schema_version: Literal[1] = 1
    type: Literal["pose.result"] = "pose.result"
    session_id: str
    captured_at_ms: int
    joints_3d: dict[str, Joint3D]
    angles: list[AngleMetric]
    primary_angle_degrees: float | None
    rep_count: int
    movement_state: str
    form_quality: Literal["good", "check", "unknown"]
    metadata: ResultMetadata


class ProjectionCalibration(Contract):
    calibration_id: str
    device_a: str
    device_b: str
    camera_matrix_a: list[list[float]]
    distortion_a: list[float]
    camera_matrix_b: list[list[float]]
    distortion_b: list[float]
    rotation_a_to_b: list[list[float]]
    translation_a_to_b: list[float]
    world_transform: list[list[float]] = Field(
        default_factory=lambda: [
            [1.0, 0.0, 0.0, 0.0],
            [0.0, 1.0, 0.0, 0.0],
            [0.0, 0.0, 1.0, 0.0],
            [0.0, 0.0, 0.0, 1.0],
        ]
    )
    units: str = "centimeters"
    reprojection_error: float | None = Field(default=None, ge=0.0)


class CalibrationStatus(Contract):
    session_id: str
    captures_by_device: dict[str, int]
    complete_pairs: int
    calibrated: bool
    calibration_id: str | None = None
    reprojection_error: float | None = None


class FinalizeCalibrationRequest(Contract):
    device_a: str
    device_b: str
    checkerboard_columns: int = Field(default=9, ge=3, le=30)
    checkerboard_rows: int = Field(default=6, ge=3, le=30)
    square_size: float = Field(default=2.5, gt=0.0)
    minimum_pairs: int = Field(default=10, ge=3, le=100)


class RealtimeFeedbackRequest(Contract):
    exercise: str
    primary_angle_degrees: float | None
    rep_count: int
    movement_state: str
    form_quality: str
    language: str = "English"


class SessionSummaryRequest(Contract):
    exercise: str
    total_reps: int
    duration_seconds: int
    angle_min: float | None = None
    angle_max: float | None = None
    form_notes: list[str] = Field(default_factory=list)
    language: str = "English"


class AiResponse(Contract):
    text: str
    provider: str
    model: str

File: web-frontend/src/lib/contracts.ts
export type ConnectionConfig = {
  backendUrl: string;
  sessionId: string;
};

export type ConnectionPhase =
  | "idle"
  | "connecting"
  | "connected"
  | "reconnecting"
  | "closed"
  | "error";

export type PoseResult = {
  schema_version: 1;
  type: "pose.result";
  session_id: string;
  captured_at_ms: number;
  joints_3d: Record<string, Joint3D>;
  angles: AngleMetric[];
  primary_angle_degrees: number | null;
  rep_count: number;
  movement_state: string;
  form_quality: "good" | "check" | "unknown";
  form_feedback?: string | null;
  metadata: ResultMetadata;
};

export type Joint3D = {
  x: number;
  y: number;
  z: number;
  confidence: number;
  observations: number;
};

export type AngleMetric = {
  name: string;
  degrees: number;
  joint_ids: [number, number, number];
};

export type ResultMetadata = {
  coordinate_system: string;
  units: string;
  calibration_id: string;
  reprojection_error: number | null;
  source_frame_ids: Record<string, number>;
  source_timestamps_ms: Record<string, number>;
  pairing_delta_ms: number;
  processing_time_ms: number;
};

export type SessionStatus = {
  session_id: string;
  exercise: string;
  device_ids: string[];
  calibrated: boolean;
  calibration_reprojection_error: number | null;
  latest_result_at: string | null;
  started_at: string;
  ended_at: string | null;
  expires_at: string;
};

export type AiResponse = { text: string; provider: string; model: string };

export type SessionSummary = {
  session_id: string;
  exercise: string;
  duration_seconds: number;
  total_reps: number;
  angle_min: number | null;
  angle_max: number | null;
  ai_summary: string | null;
  started_at: string;
  ended_at: string | null;
};

export type SessionResults = {
  session_id: string;
  results: PoseResult[];
};

export type SocketError = {
  schema_version: 1;
  type: "error";
  code: string;
  message: string;
  request_id: string | null;
};

export type SessionSocketStatus = {
  schema_version: 1;
  type: "session.status";
  session_id: string;
  status: string;
  connected_devices: number;
  calibrated: boolean;
};

export type PosePreview = {
  schema_version: 1;
  type: "pose.preview";
  session_id: string;
  device_id: string;
  frame_id: number;
  captured_at_ms: number;
  image: { width: number; height: number; rotation_degrees: number; mirrored: boolean };
  keypoints: Array<{ id: number; x: number; y: number; confidence: number }>;
};

export type SocketMessage =
  | PoseResult
  | SocketError
  | SessionSocketStatus
  | PosePreview
  | { schema_version: 1; type: "pong" };

export type AngleSample = {
  capturedAt: number;
  elapsedSeconds: number;
  angle: number;
  rep: number;
  quality: PoseResult["form_quality"];
};
